renderdoc/driver/vulkan/vk_counters.cpp

```diff
@@ -130,11 +239,40 @@ vector<CounterResult> VulkanReplay::FetchCounters(const vector<uint32_t> &counte
 
   VkDevice dev = m_pDriver->GetDev();
 
-  VkQueryPoolCreateInfo poolCreateInfo = {
+  VkQueryPoolCreateInfo timeStampPoolCreateInfo = {
       VK_STRUCTURE_TYPE_QUERY_POOL_CREATE_INFO, NULL, 0, VK_QUERY_TYPE_TIMESTAMP, maxEID * 2, 0};
 
-  VkQueryPool pool;
-  VkResult vkr = ObjDisp(dev)->CreateQueryPool(Unwrap(dev), &poolCreateInfo, NULL, &pool);
+  VkQueryPoolCreateInfo occlusionPoolCreateInfo = {
+      VK_STRUCTURE_TYPE_QUERY_POOL_CREATE_INFO, NULL, 0, VK_QUERY_TYPE_OCCLUSION, maxEID, 0};
+
+  VkQueryPipelineStatisticFlags pipeStatsFlags =
+      VK_QUERY_PIPELINE_STATISTIC_INPUT_ASSEMBLY_VERTICES_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_INPUT_ASSEMBLY_PRIMITIVES_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_VERTEX_SHADER_INVOCATIONS_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_INVOCATIONS_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_GEOMETRY_SHADER_PRIMITIVES_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_CLIPPING_INVOCATIONS_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_CLIPPING_PRIMITIVES_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_FRAGMENT_SHADER_INVOCATIONS_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_TESSELLATION_CONTROL_SHADER_PATCHES_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_TESSELLATION_EVALUATION_SHADER_INVOCATIONS_BIT |
+      VK_QUERY_PIPELINE_STATISTIC_COMPUTE_SHADER_INVOCATIONS_BIT;
+
+  VkQueryPoolCreateInfo pipeStatsPoolCreateInfo = {
+      VK_STRUCTURE_TYPE_QUERY_POOL_CREATE_INFO, NULL,   0,
+      VK_QUERY_TYPE_PIPELINE_STATISTICS,        maxEID, pipeStatsFlags};
+
+  VkQueryPool timeStampPool;
+  VkResult vkr =
+      ObjDisp(dev)->CreateQueryPool(Unwrap(dev), &timeStampPoolCreateInfo, NULL, &timeStampPool);
+  RDCASSERTEQUAL(vkr, VK_SUCCESS);
+
+  VkQueryPool occlusionPool;
+  vkr = ObjDisp(dev)->CreateQueryPool(Unwrap(dev), &occlusionPoolCreateInfo, NULL, &occlusionPool);
+  RDCASSERTEQUAL(vkr, VK_SUCCESS);
+
+  VkQueryPool pipeStatsPool;
+  vkr = ObjDisp(dev)->CreateQueryPool(Unwrap(dev), &pipeStatsPoolCreateInfo, NULL, &pipeStatsPool);
```

Comment:
You need to check for and enable the device feature `pipelineStatisticsQuery` to be able to create a pipeline statistics query pool.

Also although occlusion queries are always supported, you need to check for and enable the `occlusionQueryPrecise` feature to pass the `VK_QUERY_CONTROL_PRECISE_BIT` flag. Without that flag, the values are probably useless for our purposes so I'd say you might as well consider it an enable bit.

My suggestion would be to check these device features both before reporting the available queries in `EnumerateCounters`, and then here you can check them again and conditionally create the pool - then in the callback if the pool is NULL, just skip that query.

You'll find existing code that tries to enable any features that the replay might want (and prints an appropriate warning about what features won't be working) in `WrappedVulkan::Serialise_vkCreateDevice`.